### DISCUSSION TITLE
Feedback about the Tags feature

### DISCUSSION THREAD

**Discussion by michael-s-molina** [apache/superset#25918] [General]
Hi folks. We reviewed and quickly tested the Tags feature, and would like to provide some feedback before it’s officially released. We used the `master` version of Superset and the information available from the [Live Demo: Content Tagging in Apache Superset
](https://www.youtube.com/watch?v=tWmDlSbNeQA) for the review.

Here are some of the improvements we suggest:

**1 - Only admins can add tags**

Would it be possible to eliminate this restriction and use the permissions module for this purpose? It seems more consistent with the rest of our resources and it would give organizations more flexibility to configure who can add tags and even if tags can be added or just imported from an external system.

**2 - Displaying tags**

We believe it would be really important to display the tags when viewing an asset and not only in the list views. For example, when a user is editing a chart or viewing a dashboard. It seems like the metadata bar would be a good candidate for this.

Additionally, we foresee some assets (dashboards especially) having 10+ tags, so the UI/UX should allow users to view all these tags easily. 

**3 - Configure which assets can have tags**

Currently dashboards, charts and saved queries can have tags. We believe not everyone might want to enable tagging across all of these assets (for example we want to leverage tagging for dashboards and charts but not saved queries). Being able to configure which assets allow tagging seems like a nice to have configuration.

**4 - API endpoints**

We identified some API endpoint inconsistencies when comparing the tags API with other application endpoints and REST API best practices:
- We shouldn't have verbs when describing a resource like `/api/v1/tag/get_objects`
- Bulk create operations shouldn't be a separate endpoint like  `/api/v1/tag/bulk_create` but a naturally supported operation of `/api/v1/tag` which can accept one or more tags
- The `/api/v1/tag/favorite_status` description reads "Check favorited dashboards for current user" which seems to conflict with `/api/v1/dashboard/favorite_status` which also has the same description

**5 - Add tag action**

Shouldn't the Add Tag be a button similarly to the other actions?

<img width="562" alt="Screenshot 2023-11-07 at 14 10 47" src="https://github.com/apache/superset/assets/70410625/35524140-5c7c-4982-83ba-11dd3c670bd0">

We also noticed that you can add a tag to multiple resources but can't remove a tag from multiple resources using bulk selection.

**6 - Lists consistency**

Can we make the order of the Tags column consistent between lists? Here are the dashboards and charts lists for comparison:

<img width="1766" alt="Screenshot 2023-11-07 at 14 11 44" src="https://github.com/apache/superset/assets/70410625/3ff768b6-9d47-43f8-b44f-fd05d4ba4016">
<img width="1759" alt="Screenshot 2023-11-07 at 14 28 10" src="https://github.com/apache/superset/assets/70410625/98b1915c-3f5d-4611-b996-de1c3ec1b45e">

It looks like a good place would be before the Actions column. We could also change the Modified by and Created by columns to use avatars like the Owners column to give more space to the new column and improve consistency.

**7 - Tags page UI inconsistencies**

<img width="1769" alt="Screenshot 2023-11-07 at 14 35 00" src="https://github.com/apache/superset/assets/70410625/18851d1c-85fb-41a8-8866-00ce4f050583">

When accessing a tag's page, there are some inconsistencies when comparing the list with other lists in the application:
- Font sizes don't match
- Pagination is only being shown after a certain number of results which removes the feature of knowing how many results are available
- Should the lists be collapsible in the same way as in the home page? Or maybe use tabs for the different types just like the home page does for Favorite/Mine/All?

**8 - Edit tag modal is not scalable**

The edit modal allows a user to directly associate the tags with some resources. 

<img width="597" alt="Screenshot 2023-11-07 at 14 36 38" src="https://github.com/apache/superset/assets/70410625/8aeac5c2-4797-4179-bb3e-7ec24a1b5314">

The problem with this approach is that it's not scalable:
- The number of resources that accept tags will increase over time. Datasets, databases, alerts and reports, etc are some of the resources that may be tagged in the future.
- We can have thousands of dashboards and charts, and in this case, the select component is not a good way to implement this type of interaction.

We propose only including the tag name and description fields in this modal and do these types of associations in the resource lists where bulk selection is available.

**9 - Bugs found while reviewing the feature**

- When you add a tag, and then click to add a tag again, it retains the values of the previously added tag
- When clicking on a tag in the charts or dashboards list, it opens in a new tab breaking the SPA concept. We think the behavior should be similar to other places of the application where the default behavior is to open in the same tab but give users the option of opening in a new tab
- When you hover over a tag, it's not showing the tags description
- In the Saved Queries list, the Tags filter is appearing multiple times

**10 - Thanks**

Last but not least, thank you for taking the time to implement this feature. It's a highly requested feature that will provide a lot of value for Superset! We hope that with the improvement of our release and SIP processes, feature reviews will be more frequent and done before implementation to reduce rework.

@john-bodley, @justinpark , @michael-s-molina, @zuzana-vej

**Comment by michael-s-molina** [comment]
Pinging some folks that might be interested in this discussion: @yousoph @kasiazjc @hughhhh @rusackas @eschutho @villebro

**Reply by eschutho** [reply]
Thank you @michael-s-molina for all this great feedback!

**Reply by michael-s-molina** [reply]
It's from our whole team @john-bodley @justinpark @zuzana-vej :technologist: :technologist: :technologist: :woman_technologist:

**Comment by yousoph** [comment]
Thanks for the feedback!! 

1. Only admins can add tags
I believe there's a can_write Tag permission that can be used to give other roles tag creation permission as well, so you can customize this. 

2. Displaying tags 
Agreed that the metadata bar would be a good place for this! Definitely open to adding it there. 

3. Configuring which assets can have tags
That makes sense - perhaps a config could be useful here for this customization.

4. API endpoints
Using verbs when describing a resource: this was part of the original implementation but not opposed to changes with a breaking change release, we can discuss further. 
Bulk create operations - also agreed here, this was originally created as a new endpoint to not impact existing users that were using the POST already 
Favorite descriptions - updated here https://github.com/apache/superset/pull/25996 

6. Add tag action
We didn't feel that Add Tags should be a primary action here - it could be a secondary action potentially. In general it might be worth revisiting the bulk select action + patterns. 
For remove tag, that could be a new action here as well. 

7. List consistency 
We followed the rule of adding the Tags column before the Modified By column in the list views... though the list views aren't that consistent in general yet. I believe @villebro was looking at reordering the list views as a whole (before Tags) to make them more consistent 

8. Tags page UI 
Pagination missing might be a bug. I'm not sure that tabs would be the best way to go here, as I like the idea of being able to see multiple object types easily, but agree that this page could use some additional enhancements. 

9. Edit tag modal 
Tag creation felt like a natural place to allow users to tag things right away. In general we felt ok with the multi select pattern here since users can search the list, but of course there's also the bulk tagging option to add tags. 

10. Bugs 
Tag modal state - fix is in progress
SPA - makes sense 
Hover - yes, we will be adding it! 
Saved Queries list - thanks for reporting this one, fix coming as well.

**Comment by lightofindia1** [comment]
Hi team, it would be great if you could also support filtering charts by Tags while creating/editing Dashboard as well. Thanks!

**Comment by rajyalakshmi-stack** [comment]
Hi, can we have a multiple select option when filtering using tags. Thanks

**Reply by Sampohox** [reply]
Pls

**Comment by Quatters** [comment]
Hello everyone. It would also be great if tags will be used in import/export process. Currently, if you export some resource with tag applied on it, the tag will not be created/updated and applied to this resource. Thanks!

**Reply by rusackas** [reply]
CC @yousoph who was just discussing all the minutiae of what should/shouldn't be included in import/export.

**Comment by nik-baj** [comment]
Hello @michael-s-molina , Hello Everyone,
The TAGGING feature is just aces! Currently, tags can be added only the charts and dashboard (please correct if I am mistaken); is there a plan to extend tags to datasets and SQL Lab queries too?

**Reply by Quatters** [reply]
Tagging alerts & reports would also be useful.
